app/src/split/serial/central.c

```diff
@@ -0,0 +1,190 @@
+            if (changed_positions[i] & BIT(j)) {
+                uint32_t position = (i * 8) + j;
+                bool pressed = position_state[i] & BIT(j);
+                struct zmk_position_state_changed ev = {
+                    .position = position, .state = pressed, .timestamp = k_uptime_get()};
+
+                if (position > ZMK_KEYMAP_LEN) {
+                    LOG_WRN("Invalid position: %u", position);
+                    continue;
+                }
+
+                k_msgq_put(&peripheral_event_msgq, &ev, K_NO_WAIT);
+                k_work_submit(&peripheral_event_work);
+            }
+        }
+    }
+
+    return 0;
+}
+
+static char *alloc_position_state_buffer() {
+    char *block_ptr = NULL;
+    if (k_mem_slab_alloc(&split_memory_slab, (void **)&block_ptr, K_NO_WAIT) == 0) {
+        memset(block_ptr, 0, SPLIT_DATA_LEN);
+    } else {
+        LOG_WRN("Memory allocation time-out");
+    }
+    return block_ptr;
+}
+
+
+static void free_position_state_buffer(char *block_ptr) {
+    k_mem_slab_free(&split_memory_slab, (void **)&block_ptr);
+}
+
+
+static void uart_callback(const struct device *dev, struct uart_event *evt, void *user_data) {
+    char *buf = NULL;
+
+    switch (evt->type) {
+
+        case UART_RX_STOPPED:
+            LOG_DBG("UART device:%s rx stopped", serial_dev->name);
+            break;
+
+        case UART_RX_BUF_REQUEST:
+            LOG_DBG("UART device:%s rx extra buf req", serial_dev->name);
+            buf = alloc_position_state_buffer();
+            if (NULL != buf) {
+                int ret = uart_rx_buf_rsp(serial_dev, buf, sizeof(split_data_t));
+                if (0 != ret) {
+                    LOG_WRN("UART device:%s rx extra buf req add failed: %d", serial_dev->name, ret);
+                    free_position_state_buffer(buf);
+                }
+            }
+            break;
+
+        case UART_RX_RDY:
+            LOG_DBG("UART device:%s rx buf ready", serial_dev->name);
+            break;
+
+        case UART_RX_BUF_RELEASED:
+            LOG_DBG("UART device:%s rx buf released", serial_dev->name);
+            split_central_notify_func(evt->data.rx_buf.buf, sizeof(split_data_t));
+            free_position_state_buffer(evt->data.rx_buf.buf);
+            break;
+
+        default:
+            LOG_DBG("UART device:%s unhandled event: %u", serial_dev->name, evt->type);
+            break;
+    };
+    return;
+}
+
+
+static void split_serial_receive_thread(void *unused, void *unused1, void *unused2) {
+    if (!device_is_ready(serial_dev)) {
+        LOG_WRN("UART device:%s not ready", serial_dev->name);
+        return;
+    }
+
+    int ret = uart_callback_set(serial_dev, uart_callback, NULL);
+    if (ret == -ENOTSUP || ret == -ENOSYS) {
+        LOG_WRN("UART device:%s ASYNC not supported", serial_dev->name);
+        return;
+    }
+
+    uart_ready = 1;
+    LOG_DBG("UART device:%s ready", serial_dev->name);
+
+    while(true) {
```

Comment:
Do we really need to hard loop here? Can't we just enable the RX here, and then in the notify callback, do more? Not familiar w/ the UART API in detail, but this seems off.